### ISSUE TITLE
Provide a Bash script for Linux that automates steps during sidechain node setup

### ISSUE THREAD

**Issue by Cermelli** [HorizenOfficial/Sidechains-SDK#152]
**HDE task**

Starting points:

- https://github.com/HorizenOfficial/Sidechains-SDK/blob/master/README.md
- https://github.com/HorizenOfficial/zend_oo/blob/sidechains_testnet/README.md

Script should:

- Automate all the steps needed to declare sidechains and run a first sidechain node
- Generate all the needed keys
- Include customisation of the initial amount to be sent to the sidechain during sidechain declaration
- Declare the sidechain on the mainchain with the correct command
- Generate sidechain configuration files
- Start a sidechain node and start forging
- Include help message with detailed description of possible usage
- Be written in Bash on Linux

Installing additional tools is allowed (e.g. jq).

**For more info:** https://hde.horizen.global/tasks
***Limit first accepted submissions

**Comment by morganblack25** [comment]
pero sera compatible también para todos los sistemas operativos  de 32 bits ?

**Comment by rosariopabst** [comment]
Por ahora zend_oo y el sidechain-SDK solo es compatible con 64 bit. Vas a empezar a desarollar el bash script/bounty?

**Comment by morganblack25** [comment]
si eso quiero realizar pero mi sistema operativo es de 32bits

**Comment by jtoownus** [comment]
Hey @morganblack25 ,
if possible, please keep discussions in English so everybody can follow along. Thank you :)

**Comment by Rifat0000** [comment]
We love u zen.... Every one know zen is best

**Comment by salvatoretrimarchi** [comment]
Yes of course, how may I help you? I'm Bash Developer

**Comment by ponta9** [comment]
I’ve got a working bash script that generates all keys and customizes a sidechain node with regtest flag at the moment(with custom amount). It starts the node and automatic forger. Just missing some good error logging and test it on testnet instead.

Does the script need to compile zend_oo and compile the sidechain app?

**Comment by cronicc** [comment]
> Does the script need to compile zend_oo and compile the sidechain app?

@ponta9 compiling zend_oo in the script is optional, but nice to have. The script should at least have the option to provide a path to zend/zen-cli, or check that they are in $PATH. A check for the latest version should also be made (`zend -version`). You could also use the precompiled binaries from https://github.com/ZencashOfficial/zend_oo/releases/latest, but in that case authenticity of the binaries should be checked using gpg.

Sidechains-SDK/simpleapp should be compiled by the script, but only if the jars don't yet exist or are outdated.

**Comment by rosariopabst** [comment]
> I’ve got a working bash script that generates all keys and customizes a sidechain node with regtest flag at the moment(with custom amount). It starts the node and automatic forger. Just missing some good error logging and test it on testnet instead.
> 
> Does the script need to compile zend_oo and compile the sidechain app?

@ponta9 sounds great. marking this as a WIP. What's your ETC?

**Comment by ponta9** [comment]
> > I’ve got a working bash script that generates all keys and customizes a sidechain node with regtest flag at the moment(with custom amount). It starts the node and automatic forger. Just missing some good error logging and test it on testnet instead.
> > Does the script need to compile zend_oo and compile the sidechain app?
> 
> @ponta9 sounds great. marking this as a WIP. What's your ETC?
@rosariopabst
Had to implement some error handling. Also had to change some code to fit testnet as well as regnet. But that and logging should be done now. The only thing I have left is compiling the SDK if it's outdated and checking zend -version.

> @ponta9 compiling zend_oo in the script is optional, but nice to have. The script should at least have the option to provide a path to zend/zen-cli, or check that they are in $PATH. A check for the latest version should also be made (`zend -version`). You could also use the precompiled binaries from https://github.com/ZencashOfficial/zend_oo/releases/latest, but in that case authenticity of the binaries should be checked using gpg.
> 
> Sidechains-SDK/simpleapp should be compiled by the script, but only if the jars don't yet exist or are outdated.

@cronicc ok but right now I'm just going to check that zend and zen-cli are in the provided path.

Just got some more questions

1. **How can I check that its the latest SDK jars versions in the best way?**
    
    Should I pull that from some URL?  Check some build file on drive? Or should this be hardcoded? 
    
2. **How can I check that its the latest Zend version?**

   Should I check that version from some URL?. I could check the tag version on GitHub here.

   Or should this be hardcoded variable?

3. **Check for minimal Java and Maven version?**

   Because the script should build maven it needs to check java version and maven version

   Should this be hardcoded? In that case what minimum version of java 8 and what minimum version of maven?

   Or should it be check from some URL?

4. **Is it ok to parse a pure JSON config as a template?**

   The SDK/Sidechainapp uses HOCON for the config and there are no command line parsers that I could find without external dependencies. Right now I'm using a hardcoded config in the script.(my config has some pure JSON parts right now)

   But because valid JSON is valid HOCON and I'm already using JQ. The best way to solve this problem for me would be to use a pure JSON template that is updated each release(that way if some small stuff changes in the config it won't break the script directly). Also talked a bit about this(right at the end) in my report for the task [Create a new sidechain](https://github.com/HorizenOfficial/Sidechains-SDK/issues/151)

**Comment by alexfabian0** [comment]
Quero fazer parte disto win 64 bites a todo vapor

**Comment by cronicc** [comment]
@ponta9 
> 1. **How can I check that its the latest SDK jars versions in the best way?**
Going forward we will be doing tagged releases of the SDK, you can use the GH API to check that the latest tag matches the version in pom.xml and versions of the jars.
> 2. **How can I check that its the latest Zend version?**
You can check the latest tag on https://github.com/HorizenOfficial/zend_oo against the output from `zend -version`
> 3. **Check for minimal Java and Maven version?**
Any java >= 8 and maven >= 3 should work.
> 4. **Is it ok to parse a pure JSON config as a template?**
I've not tried if simpleapp works with a pure JSON config, if it does it's certainly ok to use a JSON template. If not you might have to look at converting JSON to HOCON, if there are no tools to do this you might have to use sed/awk etc.

**Comment by ponta9** [comment]
@cronicc Well Im using githubs api to check the zend_oo version tag. But because the SDK releases are not "proper" in GitHub api eyes i got a error response. Here is the calls im making
[zend_oo call](https://api.github.com/repos/HorizenOfficial/zend_oo/releases/latest)
[SDK call](https://api.github.com/repos/HorizenOfficial/Sidechains-SDK/releases/latest)

**Comment by cronicc** [comment]
Made the 0.2.2 tag a release @ponta9

**Comment by angievi** [comment]
Hello! @ponta9, your script is being reviewed by @zennermac and will also get a couple of more reviews in the meantime. We will come back to you with any feedback and about your bounty in the next following days. thank you!

**Comment by zennermac** [comment]
Hi @ponta9 

Firstly thanks for the great work you have done on this so far.
And also thanks for taking the time to chat verbally about this yesterday.

The script works well but as we chatted about I have some suggestions.

We should give the user the option to not automatically start the forging process as some users may see value in only having the script generate the settings file and build the jars.

It would also be nice to be able to prompt the user for paths such as custom storage locations, custom path to the zend binary etc. And even set the particular class name to use.

I know we also talked about custom jar file names but lets see if we can have the above addressed first.
